### PR TITLE
New conversion tool

### LINKED ISSUE #245: What about vim, zsh, bash, ls, grep, dircolors, tmux, etc?
It would be absolutely cool to extend this project to many Linux tools and create something like the Dracula project (https://draculatheme.com) but for iTerm2 themes.

### PR DESCRIPTION
**Pull request by tehaleph**
Almost two years later, I'm back with a new version of the conversion tool 👾

For context, below is comment from the original PR #256:
>This PR changes the logic behind theme generators. As we discussed in https://github.com/mbadolato/iTerm2-Color-Schemes/issues/245, it would be nice to be able to more easily add new formats for other programs. To facilitate this process, I removed script-based generators and rewrote it to all-in-one tool using jinja2 template engine.
Now you can add new programs simply by adding templates to specific folder. I'll update readme a bit later with information on how to do this.
I migrated all but two generators: standard macos terminal and tilda. Unlike other templates, they do not use hex or rgb color representation, but use their own data format. I'll migrate them later, anyway themes for them have not been updated for a long time. There is also a generator for hterm, but themes also does not exist, so I did not migrate it. All other generators (screenshots, html, md) are not affected by this PR.

The difference between this PR and the previous one:
- Added tilda themes generator
- Added all new schemas
- Updated all dependencies
- Added instructions for contributors

It's still not very clear what to do with generators for hterm and terminal, but I think we'll find some solution later.

Also, after merging this PR, I plan to refactor a few templates and themes. I did not add this to the current PR, as there will be a lot of generated files.

Resolve #245

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -86,12 +94,61 @@ Have a great iTerm theme? Send it to me via a Pull Request!
 2. Put your theme file into `/schemes/`
     - `mv <your-itermcolors-file> schemes/`
 3. Generate other formats for your theme using the `update_all.py` script.
-    - `cd tools/ && python3 update_all.py`
-4. Get a screenshot of your theme using the `screenshotTable.sh` script and ImageMagick. **For screenshot consistency, please have your font set to 13pt Monaco and no transparency on the window.**
+    - `cd tools/ && python3 gen.py` OR
+    - `cd tools/ && ./gen.py`
+4. If you only want to generate files for your theme, you can specify this with the `-s` flag.
+    - `./gen.py -s Dracula`
```
Comment (jdhmtl) [inline README.md:100]: Oh, that's a nice touch :)

File: README.md [hunk 2]
```diff
@@ -86,12 +94,61 @@ Have a great iTerm theme? Send it to me via a Pull Request!
...
+    - {{ Backgroun_Color.guint16 }} for guint16 representation
+
+  Also you have access to this metadata fields:
+    - {{ Guint16_Palette }} with a string containing all ansi colors as guint16 values
+    - {{ Dark_Theme }} which contains a sign that the theme is dark
+```
+3. If you need a new value type for color, add it too `tools/converter.py`
+4. Put your template file into `tool/templates`. A folder with schemas will be created based on the filename. And the file extension will remain with all generated ones. Example: `editor.ext` file will generate schemas as `editor/scheme_name.ext`
+5. Generate all existing themes for all templates `cd tools/ && ./gen.py`. Or, if you only want to generate schemas for your template, you can use the `-t` flag.
+  - `./gen.py -t kitty`
```
Comment (jdhmtl) [inline README.md:150]: Again, this level of granularity is a nice add.

File: tools/templates/electerm.txt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+terminal:blue=#{{ Ansi_4_Color.hex }}
+terminal:magenta=#{{ Ansi_5_Color.hex }}
+terminal:cyan=#{{ Ansi_6_Color.hex }}
+terminal:white=#{{ Ansi_7_Color.hex }}
+terminal:brightBlack=#{{ Ansi_8_Color.hex }}
+terminal:brightRed=#{{ Ansi_9_Color.hex }}
+terminal:brightGreen=#{{ Ansi_10_Color.hex }}
+terminal:brightYellow=#{{ Ansi_11_Color.hex }}
+terminal:brightMagenta=#{{ Ansi_12_Color.hex }}
+terminal:brightPurple=#{{ Ansi_13_Color.hex }}
```
Comment (jdhmtl) [inline tools/templates/electerm.txt:48]: See #380. These should be `brightBlue` and `brightMagenta` respectively.

File: tools/converter.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            colors_dict['Guint16_Palette'] = '{%s}' % ', '.join(guint16_pallete)
+            colors_dict['Dark_Theme'] = self.detect_dark_theme(int(colors_dict['Background_Color']['hex'], 16))
+
+        f.close()
+
+        # xrdb files were used to generate themes, now it is not needed
+        # but in case someone uses them for other purposes, these files continue to be saved
+        if not os.path.isfile(xrdb_path):
+            self.generate_xrdb_file(xrdb_path, colors_dict)
+
```
Comment (jdhmtl) [inline tools/converter.py:100]: I think that's probably a good move, but is there any reason this couldn't also be moved to a template? As written, couldn't calling `python3 gen.py -t foo` generate themes for `foo` and `xrdb` formats?

**Review by jdhmtl — COMMENTED** [review]
Really nice work. This is _so much_ cleaner and more consistent than the existing scripts, removing both duplication and individual quirks.

**Comment by tehaleph** [comment]
> Really nice work. This is _so much_ cleaner and more consistent than the existing scripts, removing both duplication and individual quirks.

Thanks! I'll add changes from #380 and I think you're right about the xrdb files – they should be just another template.

**Comment by tehaleph** [comment]
I think I'm done with this.

Also now I understand how to generate schemas for terminal app with plist but whatever I'm trying to do is slightly different due to internal calculation so I'll create another PR with terminal and chromessh templates later.

**Comment by jdhmtl** [comment]
@tehaleph Gave it another look, everything LGTM. Is this ready to go?

**Comment by tehaleph** [comment]
@jdhmtl Yeah, let's merge. And after that I'll make a PRs with the correction of some templates.

**Comment by mbadolato** [comment]
🎉

**Pull request merged**